dev_tools/notebooks/isolated_notebook_test.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# ========================== ISOLATED NOTEBOOK TESTS ============================================
+#
+# In these tests are only changed notebooks are tested. It is assumed that notebooks install cirq
+# conditionally if they can't import cirq. This installation path is the main focus and it is
+# excercised in an isolated virtual environment for each notebook. This is also the path that is
+# tested in the devsite workflows, these tests meant to provide earlier feedback.
+
+import functools
+import glob
+import os
+import subprocess
+import sys
+import warnings
+from typing import Set
+
+import pytest
+from filelock import FileLock
+
+from dev_tools import shell_tools
+from dev_tools.env_tools import create_virtual_env
+
+# these notebooks rely on features that are not released yet
+# after every release we should raise a PR and empty out this list
+# note that these notebooks are still tested in dev_tools/notebook_test.py
+NOTEBOOKS_DEPENDING_ON_UNRELEASED_FEATURES = [
+    'docs/characterization/*.ipynb',
+]
+
+# By default all notebooks should be tested, however, this list contains exceptions to the rule
+# please always add a reason for skipping.
+SKIP_NOTEBOOKS = [
+    # skipping vendor notebooks as we don't have auth sorted out
+    "**/google/*.ipynb",
+    "**/pasqal/*.ipynb",
+    "**/aqt/*.ipynb",
+    # skipping fidelity estimation due to
+    # https://github.com/quantumlib/Cirq/issues/3502
+    "examples/*fidelity*",
+] + NOTEBOOKS_DEPENDING_ON_UNRELEASED_FEATURES
+
+# As these notebooks run in an isolated env, we want to minimize dependencies that are
+# installed. We assume colab packages (feel free to add dependencies here that appear in colab, as
+# needed by the notebooks) exist. These packages are installed into a base environment as a starting
+# point, that is then cloned to a separate folder for each test.
+PACKAGES = [
+    # for running the notebooks
+    "papermill",
+    "jupyter",
+    "virtualenv-clone",
+    # assumed to be part of colab
+    "seaborn~=0.11.1",
+    # https://github.com/nteract/papermill/issues/519
+    'ipykernel==5.3.4',
+]
+
+
+# TODO(3577): extract these out to common utilities when we rewrite bash scripts in python
+def _find_base_revision():
+    for rev in ['upstream/master', 'origin/master', 'master']:
+        try:
+            out = subprocess.check_output(f'git cat-file -t {rev}'.split())
```

Comment:
Thanks for the pointer - in fact `subprocess.run` is the latest addition that does what we need here.

Reply:
sorry, that must have been what I meant! https://docs.python.org/3/library/subprocess.html 

> The recommended approach to invoking subprocesses is to use the run() function for all use cases it can handle. For more advanced use cases, the underlying Popen interface can be used directly.